src/odemis/dataio/tiff.py

```diff
@@ -1955,4 +1718,461 @@ def read_thumbnail(filename):
+        Fetches a part of the data, for a given zoom. If the (complete) data has more
+        than two dimensions, all the extra dimensions (ie, non-spatial) are always fully
+        returned for the given part.
+        n (int): index of the image
+        z (0 <= int) : zoom level. The data returned will be with MD_PIXEL_SIZE * 2^z.
+            So 0 means to use the highest zoom, with the original pixel size. 1 will
+            return data half the width and heigh (The maximum possible value depends
+            on the data).
+        rect (4 ints): left, top, right, bottom coordinates (in px, at zoom=0) of the
+            area of interest.
+        return (tuple of tuple of DataArray): all the tiles in X&Y dimension, so that
+            the area of interest is fully covered (so the area can be larger than requested).
+            The first dimension is X, and second is Y. For example, if returning 3x7 tiles,
+            the most bottom-right tile will be accessed as ret[2][6]. For each
+            DataArray.metadata, MD_POS and MD_PIXEL_SIZE are updated appropriately
+            (if MD_POS is not present, (0,0) is used as default for the entire image, and if
+            MD_PIXEL_SIZE is not present, it will not be updated).
+        raise ValueError: if the area or z is out of range, or if the raw data is not pyramidal.
+        """
+        if not (0 <= n < len(self.content)):
+            raise ValueError("Invalid N value %d" % (n,))
+
+        # get information about how to retrieve the actual pixels from the TIFF file
+        tiff_info = self.content[n].tiff_info
+        # TODO Implement the reading of the subdata when tiff_info is a list. 
+        # It is the case when the DataArray has multiple pixelData (eg, when data has more than 2D).
+        if type(tiff_info) is list:
+            raise NotImplemented("Not implemented when DataArray has multiple pixelData")
+
+        tiff_file = tiff_info['handle']
+        tiff_file.SetDirectory(tiff_info['dir_index'])
+
+        if z != 0:
+            # get an array of offsets, one for each subimage
+            sub_ifds = tiff_file.GetField(T.TIFFTAG_SUBIFD)
+            if not sub_ifds:
+                raise ValueError("Image does not have zoom levels")
+
+            if not (0 <= z <= len(sub_ifds)):
+                raise ValueError("Invalid Z value %d" % (z,))
+
+            # set the offset of the subimage. Z=0 is the main image
+            tiff_file.SetSubDirectory(sub_ifds[z - 1])
+        
+        num_tcols = tiff_file.GetField(T.TIFFTAG_TILEWIDTH)
+        num_trows = tiff_file.GetField(T.TIFFTAG_TILELENGTH)
+
+        if not num_tcols or not num_trows:
+            raise RuntimeError("the image is not tiled")
+
+        orig_pixel_size = self.content[n].metadata.get(model.MD_PIXEL_SIZE, (1, 1))
+        # calculate the pixel size of the tile for the zoom level
+        tile_pixel_size = tuple([ps * 2 ** z for ps in orig_pixel_size])
+
+        x1, y1, x2, y2 = rect
```

Comment:
It's using rect as containing coordinates in pixels at the given zoom level, right? It should be in the coordinates of pixels at zoom = 0. (=> so you need to divide by 2**z and round it).